include/boost/math/interpolators/cubic_b_spline.hpp

```diff
@@ -0,0 +1,292 @@
+// four-times continuously differentiable, the error is of O(h^4).
+// In addition, we can differentiate the spline and obtain a good interpolant for f'.
+// The main restriction of this method is that the samples of f must be evenly spaced.
+// Interpolators for non-evenly sampled data are forthcoming.
+// Properties:
+// - s(x_j) = f(x_j)
+// - All cubic polynomials interpolated exactly
+
+#ifndef BOOST_MATH_INTERPOLATORS_CUBIC_B_SPLINE_HPP
+#define BOOST_MATH_INTERPOLATORS_CUBIC_B_SPLINE_HPP
+
+#include <limits>
+#include <cmath>
+#include <vector>
+#include <boost/math/constants/constants.hpp>
+#include <boost/math/special_functions/fpclassify.hpp>
+
+namespace boost{ namespace math{
+
+using boost::math::constants::third;
+using boost::math::constants::half;
+
+template<class Real>
+Real b3_spline(Real x)
+{
+    auto sixth = half<Real>()*third<Real>();
+    using std::abs;
+    auto absx = abs(x);
+    if (absx < 1)
+    {
+        auto y = 2 - absx;
+        auto z = 1 - absx;
+        return sixth*(y*y*y - 4*z*z*z);
+    }
+    if (absx < 2)
+    {
+        auto y = 2 - absx;
+        return sixth*y*y*y;
+    }
+    return (Real) 0;
+}
+
+template<class Real>
+Real b3_spline_prime(Real x)
+{
+    if ( x < 0 )
+    {
+        return -b3_spline_prime(-x);
+    }
+
+    if (x < 1)
+    {
+        return x*(3*half<Real>()*x - 2);
+    }
+    if (x < 2)
+    {
+        return -half<Real>()*(2 - x)*(2 - x);
+    }
+    return (Real) 0;
+}
+
+template <class Real>
+class cubic_b_spline
+{
+public:
+    // If you don't know the value of the derivative at the endpoints, leave them as nans and the routine will estimate them.
+    // f[0] = f(a), f[length -1] = b, step_size = (b - a)/(length -1).
+    cubic_b_spline(const Real* const f, size_t length, Real left_endpoint, Real step_size,
+                   Real left_endpoint_derivative = std::numeric_limits<Real>::quiet_NaN(),
+                   Real right_endpoint_derivative = std::numeric_limits<Real>::quiet_NaN());
+
+    Real interpolate_at(Real x) const;
+
+    Real interpolate_derivative(Real x) const;
+
+private:
+    std::vector<Real> m_beta;
+    Real m_h_inv;
+    Real m_a;
+    Real m_avg;
+
+};
+
+
+template<class Real>
+cubic_b_spline<Real>::cubic_b_spline(const Real* const f, size_t length, Real left_endpoint, Real step_size,
+                                     Real left_endpoint_derivative, Real right_endpoint_derivative) : m_a(left_endpoint), m_avg(0)
+{
+    if (length < 5)
+    {
+        if (boost::math::isnan(left_endpoint_derivative) || boost::math::isnan(right_endpoint_derivative))
+        {
+            throw std::logic_error("Interpolation using a cubic b spline with derivatives estimated at the endpoints requires at least 5 points.\n");
+        }
+        if (length < 3)
+        {
+            throw std::logic_error("Interpolation using a cubic b spline requires at least 3 points.\n");
+        }
+    }
+
+    if (boost::math::isnan(left_endpoint))
```

Comment:
Added a check to see that the right endpoint doesn't overflow the requested type. If the left endpoint `= -std::numeric_limits<Real>::max()`, there should be no problem.